trin-execution/src/evm/execution_context.rs

```diff
@@ -0,0 +1,291 @@
+        self.evm
+            .db_mut()
+            .merge_transitions(BundleRetention::PlainState);
+        let state_bundle = self.evm.db_mut().take_bundle();
+        self.evm.db_mut().database.commit_bundle(state_bundle)?;
+        stop_timer(timer);
+
+        let timer = start_timer_vec(&BLOCK_PROCESSING_TIMES, &["get_root_with_trie_diff"]);
+        let RootWithTrieDiff {
+            root,
+            trie_diff: changed_nodes,
+        } = self
+            .evm
+            .db_mut()
+            .database
+            .trie
+            .lock()
+            .root_hash_with_changed_nodes()?;
+        if root != expected_state_root {
+            panic!(
+                "State root doesn't match! Irreversible! Block number: {}",
+                self.block_number()
+            )
+        }
+        stop_timer(timer);
+
+        Ok(RootWithTrieDiff {
+            root,
+            trie_diff: changed_nodes,
+        })
+    }
+
+    pub fn execute_block(&mut self, block: &ProcessedBlock) -> anyhow::Result<()> {
+        info!("State EVM processing block {}", block.header.number);
+
+        self.manage_block_hash_serve_window(block)?;
+
+        let execute_block_timer = start_timer_vec(&BLOCK_PROCESSING_TIMES, &["execute_block"]);
+
+        self.set_evm_environment_from_block(block);
+
+        // execute transactions
+        let cumulative_transaction_timer =
+            start_timer_vec(&BLOCK_PROCESSING_TIMES, &["cumulative_transaction"]);
+        for transaction in block.transactions.iter() {
+            let transaction_timer = start_timer_vec(&BLOCK_PROCESSING_TIMES, &["transaction"]);
+            let evm_result = self.execute_transaction(transaction)?;
+            self.cumulative_gas_used += evm_result.result.gas_used();
+            self.commit(evm_result.state);
+            stop_timer(transaction_timer);
+        }
+        stop_timer(cumulative_transaction_timer);
+
+        // update beneficiary
+        let timer = start_timer_vec(&BLOCK_PROCESSING_TIMES, &["update_beneficiary"]);
+        let _ = self.increment_balances(get_block_reward(block));
+
+        // check if dao fork, if it is drain accounts and transfer it to beneficiary
+        if block.header.number == get_spec_block_number(SpecId::DAO_FORK) {
+            self.process_dao_fork()?;
+        }
+        self.cumulative_gas_expected += block.header.gas_used.to::<u64>();
+        stop_timer(timer);
+        stop_timer(execute_block_timer);
+        set_int_gauge_vec(&BLOCK_HEIGHT, block.header.number as i64, &[]);
+        Ok(())
+    }
+
+    fn execute_transaction(
+        &mut self,
+        tx: &TransactionsWithSender,
+    ) -> anyhow::Result<ResultAndState> {
+        self.set_transaction_evm_context(tx);
+        let block_number = self.block_number();
+
+        let timer = start_timer_vec(&TRANSACTION_PROCESSING_TIMES, &["transact"]);
+
+        let result = if self.block_to_trace.should_trace(block_number) {
```

Comment:
Traces are very expensive as the logging to disk is massive. The use case is mainly debugging I am not aware of another. 

Also if we did that everything would need to be generic due to that.

So unless it is really needed I would like to avoid it.

So I would like to avoid it in this PR, but if you think you can do it without too much cost or complexity to justify a niche debugging feature go for it.